library/alloc/src/collections/btree/node/node_ptr.rs

```diff
@@ -0,0 +1,131 @@
+use super::{InternalNode, LeafNode};
+use crate::alloc::{AllocRef, Global, Layout};
+use crate::boxed::Box;
+use core::ptr::NonNull;
+
+/// A managed, non-null, owned pointer to a node. It points to either an owned
+/// `LeafNode<K, V>` or to the leaf portion of an owned `InternalNode<K, V>`.
+///
+/// However, `BoxedNode` contains no information as to which of the two types
+/// of nodes it actually contains, and, partially due to this lack of information,
+/// has no destructor.
+pub(super) struct BoxedNode<K, V> {
+    ptr: NonNull<LeafNode<K, V>>,
+}
+
+/// A managed, non-null pointer to a node that knows the node's height. It
+/// nodels the type:
+/// ```
+///     use core::num::NonZeroUsize;
+///     use core::ptr::NonNull;
+///     # #[allow(dead_code)] struct LeafNode<K, V>(K, V);
+///     # #[allow(dead_code)] struct InternalNode<K, V>(K, V);
+///     # #[allow(dead_code)]
+///     enum UnboxedNode<K, V> {
+///         Leaf { ptr: NonNull<LeafNode<K, V>> },
+///         Internal { height: NonZeroUsize, ptr: NonNull<InternalNode<K, V>> },
+///     }
+/// ```
+/// by way of a single pointer field and casts, because that is easier for
+/// gdb_providers.py, even more so because it is very similar to `BoxedNode`.
+///
+/// However, `UnboxedNode` contains no information as to whether it owns or
+/// shares the node and has no implicit destructor.
+pub(super) struct UnboxedNode<K, V> {
```

Comment:
I'm not sure why this is called "Unboxed" -- it seems to be equivalent to a NodeRef just without some of the type parameters?

Reply:
I'm not sure why `BoxedNode` is called what it is… it's boxed in the alloc::boxed sense of the word (though it doesn't have to be - the first leaf can in theory be integrated into `struct BTreeMap` or other wrappers) but in my mind the purpose is to squeeze it into a small box. I would probably call them (`BoxedNode` and the newcomer) `PackedNodePtr` and `NodePtr`, if they came in out of the blue.